src/main/java/com/nimbusds/jose/crypto/ECDH1PUDecrypter.java

```diff
@@ -0,0 +1,286 @@
+	 *
+	 * @param privateKey     The private EC key. Must not be {@code null}.
+	 * @param publicKey 	 The public EC key. Must not be {@code null}.
+	 * @param defCritHeaders The names of the critical header parameters
+	 *                       that are deferred to the application for
+	 *                       processing, empty set or {@code null} if none.
+	 *
+	 * @throws JOSEException If the elliptic curve is not supported.
+	 */
+	public ECDH1PUDecrypter(final ECPrivateKey privateKey,
+							final ECPublicKey publicKey,
+							final Set<String> defCritHeaders)
+		throws JOSEException {
+
+		this(privateKey, publicKey, defCritHeaders, Curve.forECParameterSpec(privateKey.getParams()));
+	}
+
+
+	/**
+	 * Creates a new Elliptic Curve Diffie-Hellman decrypter. This
+	 * constructor can also accept a private EC key located in a PKCS#11
+	 * store that doesn't expose the private key parameters (such as a
+	 * smart card or HSM).
+	 *
+	 * @param privateKey     The private EC key. Must not be {@code null}.
+	 * @param publicKey 	 The public EC key. Must not be {@code null}.
+	 * @param defCritHeaders The names of the critical header parameters
+	 *                       that are deferred to the application for
+	 *                       processing, empty set or {@code null} if none.
+	 * @param curve          The key curve. Must not be {@code null}.
+	 *
+	 * @throws JOSEException If the elliptic curve is not supported.
+	 */
+	public ECDH1PUDecrypter(final ECPrivateKey privateKey,
+							final ECPublicKey publicKey,
+                            final Set<String> defCritHeaders,
+                            final Curve curve)
+		throws JOSEException {
+
+		super(curve);
+
+		critPolicy.setDeferredCriticalHeaderParams(defCritHeaders);
+
+		this.privateKey = privateKey;
+		this.publicKey = publicKey;
+	}
+
+	/**
+	 * Returns the public EC key.
+	 *
+	 * @return The public EC key.
+	 */
+	public ECPublicKey getPublicKey() {
+
+		return publicKey;
+	}
+
+	/**
+	 * Returns the private EC key.
+	 *
+	 * @return The private EC key. Casting to
+	 *         {@link ECPrivateKey} may not be
+	 *         possible if the key is located in a PKCS#11 store that
+	 *         doesn't expose the private key parameters.
+	 */
+	public PrivateKey getPrivateKey() {
+
+		return privateKey;
+	}
+
+
+	@Override
+	public Set<Curve> supportedEllipticCurves() {
+
+		return SUPPORTED_ELLIPTIC_CURVES;
+	}
+
+
+	@Override
+	public Set<String> getProcessedCriticalHeaderParams() {
+
+		return critPolicy.getProcessedCriticalHeaderParams();
+	}
+
+
+	@Override
+	public Set<String> getDeferredCriticalHeaderParams() {
+
+		return critPolicy.getProcessedCriticalHeaderParams();
+	}
+
+
+	@Override
+	public byte[] decrypt(final JWEHeader header,
+			      final Base64URL encryptedKey,
+			      final Base64URL iv,
+			      final Base64URL cipherText,
+			      final Base64URL authTag)
+		throws JOSEException {
+
+		ECDH1PU.validateSameCurve(privateKey, publicKey);
```

Comment:
Why don't we call the same method for ephemeral keys (instead of the code below)?

Reply:
fixed